interp/src/stk_env.rs

```diff
@@ -0,0 +1,809 @@
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next.map(|node| {
+            self.next = node.next.as_deref();
+            &node.elem
+        })
+    }
+}
+
+///A Stack of HashMaps that supports scoping
+///
+///Invariants:
+///=> To Smoosh X levels down, no fork may still exist from any node amongst those
+/// X levels. So the stack must look like:
+/// ```text
+///  _
+/// |_|
+///  |
+///  _
+/// |_|
+///  |
+///  _
+/// |_|
+///  |
+/// ```
+///=> To merge, both branches must share a common fork point
+/// ```text
+///  _
+/// |_|
+///  |
+///  _     _
+/// |_|   |_|
+///  |     |
+///   \ _ /
+///    |*|
+///     |
+/// ```
+#[derive(Debug)]
+pub struct Smoosher<K: Eq + std::hash::Hash, V: Eq> {
+    head: HashMap<K, V>,       //mutable
+    tail: List<HashMap<K, V>>, //read-only
+}
+
+impl<K: Eq + std::hash::Hash, V: Eq> Smoosher<K, V> {
+    pub fn new() -> Smoosher<K, V> {
+        Smoosher {
+            head: HashMap::new(),
+            tail: List::new(),
+        }
+    }
+
+    /// If [self] and [other] share a fork point, returns a pair (depthA, depthB)
+    /// of the depth which the fork point can be found in [self] and [other], respectively.
+    /// NOTE: should be private, only public for testing!
+    pub fn shared_fork_point(&self, other: &Self) -> Option<(u64, u64)> {
+        //check head
+        if std::ptr::eq(&self.head, &other.head) {
+            return Some((0, 0));
+        } else {
+            //check tail
+            //these start at 1 b/c head was 0
+            let mut a_depth = 1;
+            let mut b_depth = 1;
+            //iterate and check all other nodes
+            for nd in self.tail.iter() {
+                b_depth = 1;
+                for nd_other in other.tail.iter() {
+                    if std::ptr::eq(nd, nd_other) {
+                        return Some((a_depth, b_depth));
+                    }
+                    b_depth += 1;
+                }
+                a_depth += 1;
+            }
+            return None; //do not share a fork point
+        }
+    }
+    // NOTE: shold be private, only public for testing!
+    pub fn drop(self) {
+        //yea
+    }
+
+    /// Returns an Option of a pointer to the highest-scoped binding to [k],
+    /// if it exists. Else None.
+    ///
+    /// # Example
+    /// ```
+    /// use interp::stk_env::Smoosher;
+    /// let mut smoosher = Smoosher::new();
+    /// smoosher.set("hi!", 1);
+    /// assert_eq!(*smoosher.get(&"hi!").unwrap(), 1);
+    /// assert_eq!(smoosher.get(&"hey"), None);
+    /// ```
+    pub fn get(&self, k: &K) -> Option<&V> {
+        //first check if it's in the highest one
+        if let Some(val) = self.head.get(k) {
+            return Some(val);
+        } else {
+            let iter = self.tail.iter();
+            for hm in iter {
+                if let Some(val) = hm.get(k) {
+                    return Some(val);
+                }
```

Comment:
You can replace this with `find_map` through the iterator

Reply:
won't `find_map` modify the iter? and also, will it return `None` if it doesn't find any value (that's desired)

Reply:
find map will consume the iter, yes, but you're doing that with the for loop anyway. And yes to the second question

Reply:
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find_map